docs/rfcs/0000-search-defaults.md

```diff
@@ -0,0 +1,246 @@
+| Network                  | 0.7    | The accuracy of the network provider is not known and so we trust it a bit less.
+| Google Play Services     | 0.7    | The fused location provider is a black box. We want to primarily use it to rank up regions that other providers returned too.
+| Mozilla Location Service | 0.9    | We use MLS for search already and therefore give it a high weight.
+| Mobile Country Code      | 0.5    | May be close to the physical location too. Maybe could be ranked higher.
+| Operator Country Code    | 0.5    | The operator code is static bound to the SIM card. It may differ from the actual region.
+
+For all regions returned by the _sources_ above we will calculate the weighted average as _score_. We will only use the region with the highest score in the next steps.
+
+If the providers return the following regions:
+
+* GPS = `US`
+* Network = `CA`
+* Google Play Services = `null`
+* Mozilla Location Service = `US`
+* Mobile Country Code = `US`
+* Operator Country Code `CA`
+
+Then we will calculate the following scores:
+
+* `US`: 1 + 0.9 + 0.5 = **2.4** (Normalized **0.6666**)
+* `CA`: 0.7 + 0.5 = **1.2** (Normalized **0.3333**)
+
+In this example we determine the _current_ region of the user to be `US` (0.6666).
+
+See the "examples" section for complete, real-world examples.
+
+### Validation
+
+In addition to querying sources for a region, there are additional sources that do not return a region themselves but let us verify the accuracy of a region.
+
+We will model this by having those _validators_ provide additional weights for regions. For example a _validator_ returning `0.5` for `DE` (meaning it's unlikely the user is in Germany) will reduce the score of all providers returning `DE`. Or if a _validator_ returns `0` for `US` then the score of all providers returning `US` will be `0`.
+
+The first _validator_ that we want to ship uses the timezone of the user. Similar to the implementation in the desktop version of Firefox ([[1]](https://searchfox.org/mozilla-central/source/toolkit/modules/Region.jsm#189-194), [[2]](https://searchfox.org/mozilla-central/source/toolkit/modules/Region.jsm#736-752)) users with a `US` region and not a `US` timezone will not get `US` assigned as _current_ or _home_ region. In our implementation this behavior can be mapped to a _validator_ that returns `0` for `US` if the user is not in a `US` timezone.
+
+### Timeouts
+
+On the very first app launch we will have no _home_ region set yet. In order to get a quick response, and to not block the search functionality, we will have to apply a stricter timeout to the region resolution. This primarily affects sources that require a network request, like MLS. Fennec used a [10 second timeout](https://searchfox.org/mozilla-esr68/rev/4b79eedfe95ef427490a16e7c96f3a0930ab0c32/mobile/android/base/java/org/mozilla/gecko/search/SearchEngineManager.java#399-400), which we will use as starting point.
+
```

Comment:
When you say timeout, are you planning to block the Search engines from being shown until this timeout is hit?

If so thats one thing I would consider rethinking, particularly when the default fallback is going to be a network request to MLS on a mobile device, on desktop we have switched that path to use the default region on startup and always treat region updates as background / async updates. Whether you have realtime updates may affect this on first run but even so I would expect on mobile we dont see the month long sessions without a restart that we can get on desktop right?

Reply:
I sketched some code for this today and came to the same conclusion: There's no need for a timeout. Like you said, we can either use the home region we already have, or the default. And then always do the lookup request  + react to region updates, which are rare.
I haven't thought about very long running sessions, but also expect them to be unlikely on mobile. Does desktop handle this? I briefly looked at region.jsm, but couldn't see anything except the failure retry.

Reply:
The patch for that is @ https://phabricator.services.mozilla.com/D91214

As was proposed doing here, we passively listen for geolocation updates and use that to update the region, but in case the user never uses any geolocation we schedule an updateTimer for once a week